algorithms/src/polycommit/sonic_pc/data_structures.rs

```diff
@@ -353,16 +353,51 @@ impl<E: PairingEngine> CommitterKey<E> {
             domain,
         })
     }
-}
 
-impl<E: PairingEngine> CommitterKey<E> {
     pub fn max_degree(&self) -> usize {
         self.max_degree
     }
 
     pub fn supported_degree(&self) -> usize {
         self.powers_of_beta_g.len() - 1
     }
+
+    pub fn union<'a, T: IntoIterator<Item = &'a Self> + Clone>(committer_keys: T) -> Self {
+        let mut union = CommitterKey::<E>{
+            powers_of_beta_g: vec![],
+            lagrange_bases_at_beta_g: BTreeMap::new(),
+            powers_of_beta_times_gamma_g: vec![],
+            shifted_powers_of_beta_g: None,
+            shifted_powers_of_beta_times_gamma_g: None,
+            enforced_degree_bounds: None,
+            max_degree: 0,
+        };
+        let mut enforced_degree_bounds = vec![];
+        let mut biggest_ck = None;
+        for ck in committer_keys {
+            if biggest_ck.is_none() {
+                biggest_ck = Some(ck);
+            } else if biggest_ck.unwrap().supported_degree() < ck.supported_degree() {
+                biggest_ck = Some(ck);
+            }
+            union.lagrange_bases_at_beta_g.append(&mut ck.lagrange_bases_at_beta_g.clone());
+            enforced_degree_bounds.append(&mut ck.enforced_degree_bounds.clone().unwrap());
```

Comment:
We should avoid cloning here as the `lagrange_bases` is quite large.

Reply:
For both the committer and verifier keys, we can either pass everything by value to avoid cloning, or create new union committer/verifier key types which reference the original committer/verifyer keys. Which setup would have your preference?